File: cineva/packages/audio_engine/native/cineva_dsp/src/biquad.c
#include "biquad.h"

#include <math.h>

#include "dsp_util.h"

void cv_biquad_init(cv_biquad *bq) {
  bq->b0 = 1.0; bq->b1 = 0.0; bq->b2 = 0.0; bq->a1 = 0.0; bq->a2 = 0.0;
  bq->tb0 = 1.0; bq->tb1 = 0.0; bq->tb2 = 0.0; bq->ta1 = 0.0; bq->ta2 = 0.0;
  bq->s1 = 0.0; bq->s2 = 0.0;
  bq->ramp = 0;
}

void cv_biquad_reset_state(cv_biquad *bq) {
  bq->s1 = 0.0;
  bq->s2 = 0.0;
}

void cv_biquad_snap(cv_biquad *bq) {
  bq->b0 = bq->tb0; bq->b1 = bq->tb1; bq->b2 = bq->tb2;
  bq->a1 = bq->ta1; bq->a2 = bq->ta2;
  bq->ramp = 0;
}

void cv_biquad_set_raw(cv_biquad *bq, double b0, double b1, double b2,
                       double a1, double a2) {
  bq->tb0 = b0; bq->tb1 = b1; bq->tb2 = b2; bq->ta1 = a1; bq->ta2 = a2;
  bq->ramp = 1;
}

void cv_biquad_set(cv_biquad *bq, int type, double freq_hz, double gain_db,
                   double q, double sample_rate) {
  double fs = sample_rate > 0.0 ? sample_rate : 48000.0;
  double f0 = cv_clamp_d(freq_hz, 10.0, fs * 0.49);
  double Q = cv_clamp_d(q, 0.1, 20.0);
  double A = pow(10.0, cv_clamp_d(gain_db, -40.0, 40.0) / 40.0);
  double w0 = 2.0 * CINEVA_PI * f0 / fs;
  double cw = cos(w0);
  double sw = sin(w0);
  double alpha = sw / (2.0 * Q);

  double b0 = 1.0, b1 = 0.0, b2 = 0.0, a0 = 1.0, a1 = 0.0, a2 = 0.0;

  switch (type) {
    case CV_BQ_LOWSHELF: {
      double sqA = 2.0 * sqrt(A) * alpha;
      b0 = A * ((A + 1.0) - (A - 1.0) * cw + sqA);
      b1 = 2.0 * A * ((A - 1.0) - (A + 1.0) * cw);
      b2 = A * ((A + 1.0) - (A - 1.0) * cw - sqA);
      a0 = (A + 1.0) + (A - 1.0) * cw + sqA;
      a1 = -2.0 * ((A - 1.0) + (A + 1.0) * cw);
      a2 = (A + 1.0) + (A - 1.0) * cw - sqA;
      break;
    }
    case CV_BQ_HIGHSHELF: {
      double sqA = 2.0 * sqrt(A) * alpha;
      b0 = A * ((A + 1.0) + (A - 1.0) * cw + sqA);
      b1 = -2.0 * A * ((A - 1.0) + (A + 1.0) * cw);
      b2 = A * ((A + 1.0) + (A - 1.0) * cw - sqA);
      a0 = (A + 1.0) - (A - 1.0) * cw + sqA;
      a1 = 2.0 * ((A - 1.0) - (A + 1.0) * cw);
      a2 = (A + 1.0) - (A - 1.0) * cw - sqA;
      break;
    }
    case CV_BQ_LOWPASS: {
      b0 = (1.0 - cw) * 0.5;
      b1 = 1.0 - cw;
      b2 = (1.0 - cw) * 0.5;
      a0 = 1.0 + alpha;
      a1 = -2.0 * cw;
      a2 = 1.0 - alpha;
      break;
    }
    case CV_BQ_HIGHPASS: {
      b0 = (1.0 + cw) * 0.5;
      b1 = -(1.0 + cw);
      b2 = (1.0 + cw) * 0.5;
      a0 = 1.0 + alpha;
      a1 = -2.0 * cw;
      a2 = 1.0 - alpha;
      break;
    }
    case CV_BQ_PEAKING:
    default: {
      b0 = 1.0 + alpha * A;
      b1 = -2.0 * cw;
      b2 = 1.0 - alpha * A;
      a0 = 1.0 + alpha / A;
      a1 = -2.0 * cw;
      a2 = 1.0 - alpha / A;
      break;
    }
  }

  if (!(a0 > 0.0) || !(a0 == a0)) a0 = 1.0;
  cv_biquad_set_raw(bq, b0 / a0, b1 / a0, b2 / a0, a1 / a0, a2 / a0);
}

File: cineva/packages/audio_engine/lib/src/dsp/biquad.dart
/// Biquad RBJ — miroir exact de `native/cineva_dsp/src/biquad.c`.
/// Transposed direct form II, coefficients interpolés (anti-zipper).
library;

import 'dart:math' as math;
import 'math_ext.dart';

const int bqLowShelf = 0;
const int bqPeaking = 1;
const int bqHighShelf = 2;
const int bqLowPass = 3;
const int bqHighPass = 4;

/// Nombre maximal de canaux gérés par le moteur (miroir de
/// `CINEVA_DSP_MAX_CHANNELS` dans `native/cineva_dsp/include/cineva_dsp.h`).
const int maxChannels = 8;

class Biquad {
  double b0 = 1, b1 = 0, b2 = 0, a1 = 0, a2 = 0;
  double _tb0 = 1, _tb1 = 0, _tb2 = 0, _ta1 = 0, _ta2 = 0;
  double _s1 = 0, _s2 = 0;
  bool _ramp = false;

  void init() {
    b0 = 1; b1 = 0; b2 = 0; a1 = 0; a2 = 0;
    _tb0 = 1; _tb1 = 0; _tb2 = 0; _ta1 = 0; _ta2 = 0;
    _s1 = 0; _s2 = 0;
    _ramp = false;
  }

  void setRaw(double tb0, double tb1, double tb2, double ta1, double ta2) {
    _tb0 = tb0; _tb1 = tb1; _tb2 = tb2; _ta1 = ta1; _ta2 = ta2;
    _ramp = true;
  }

  void snap() {
    b0 = _tb0; b1 = _tb1; b2 = _tb2; a1 = _ta1; a2 = _ta2;
    _ramp = false;
  }

  void resetState() {
    _s1 = 0;
    _s2 = 0;
  }

  void set(int type, double freqHz, double gainDb, double q, double sampleRate) {
    final double fs = sampleRate > 0 ? sampleRate : 48000;
    final double f0 = freqHz.clamp(10.0, fs * 0.49);
    final double qq = q.clamp(0.1, 20.0);
    final double gg = gainDb.clamp(-40.0, 40.0);
    final double a = math.pow(10.0, gg / 40.0).toDouble();
    final double w0 = 2 * math.pi * f0 / fs;
    final double cw = math.cos(w0);
    final double sw = math.sin(w0);
    final double alpha = sw / (2 * qq);

    double b0 = 1, b1 = 0, b2 = 0, a0 = 1, a1 = 0, a2 = 0;
    switch (type) {
      case bqLowShelf:
        final double sq = 2 * math.sqrt(a) * alpha;
        b0 = a * ((a + 1) - (a - 1) * cw + sq);
        b1 = 2 * a * ((a - 1) - (a + 1) * cw);
        b2 = a * ((a + 1) - (a - 1) * cw - sq);
        a0 = (a + 1) + (a - 1) * cw + sq;
        a1 = -2 * ((a - 1) + (a + 1) * cw);
        a2 = (a + 1) + (a - 1) * cw - sq;
      case bqHighShelf:
        final double sq = 2 * math.sqrt(a) * alpha;
        b0 = a * ((a + 1) + (a - 1) * cw + sq);
        b1 = -2 * a * ((a - 1) + (a + 1) * cw);
        b2 = a * ((a + 1) + (a - 1) * cw - sq);
        a0 = (a + 1) - (a - 1) * cw + sq;
        a1 = 2 * ((a - 1) - (a + 1) * cw);
        a2 = (a + 1) - (a - 1) * cw - sq;
      case bqLowPass:
        b0 = (1 - cw) * 0.5;
        b1 = 1 - cw;
        b2 = (1 - cw) * 0.5;
        a0 = 1 + alpha;
        a1 = -2 * cw;
        a2 = 1 - alpha;
      case bqHighPass:
        b0 = (1 + cw) * 0.5;
        b1 = -(1 + cw);
        b2 = (1 + cw) * 0.5;
        a0 = 1 + alpha;
        a1 = -2 * cw;
        a2 = 1 - alpha;
      default: // peaking
        b0 = 1 + alpha * a;
        b1 = -2 * cw;
        b2 = 1 - alpha * a;
        a0 = 1 + alpha / a;
        a1 = -2 * cw;
        a2 = 1 - alpha / a;
    }
    if (!(a0 > 0) || a0.isNaN) a0 = 1;
    setRaw(b0 / a0, b1 / a0, b2 / a0, a1 / a0, a2 / a0);
  }

  double tick(double x, double rampCoef) {
    if (_ramp) {
      b0 += (_tb0 - b0) * rampCoef;
      b1 += (_tb1 - b1) * rampCoef;
      b2 += (_tb2 - b2) * rampCoef;
      a1 += (_ta1 - a1) * rampCoef;
      a2 += (_ta2 - a2) * rampCoef;
      final double e0 = b0 - _tb0;
      final double e1 = b1 - _tb1;
      final double e2 = b2 - _tb2;
      final double e3 = a1 - _ta1;
      final double e4 = a2 - _ta2;
      if (e0 * e0 + e1 * e1 + e2 * e2 + e3 * e3 + e4 * e4 < 1e-18) {
        snap();
      }
    }
    final double y = b0 * x + _s1;
    _s1 = b1 * x - a1 * y + _s2;
    _s2 = b2 * x - a2 * y;
    return y;
  }
}

/// Utilitaires DSP partagés (miroir de dsp_util.h).
double clampD(double v, double lo, double hi) {
  if (v.isNaN) return lo;
  if (v < lo) return lo;
  if (v > hi) return hi;
  return v;
}

double sanitizeD(double v) => (v.isFinite && v > -1e30 && v < 1e30) ? v : 0.0;

double dbToLin(double db) => math.pow(10.0, db / 20.0).toDouble();

double linToDb(double lin) => 20 * log10(lin > 1e-12 ? lin : 1e-12);

double onepoleCoef(double sampleRate, double tauSeconds) {
  if (tauSeconds <= 0 || sampleRate <= 0) return 0;
  return math.exp(-1 / (sampleRate * tauSeconds));
}
